src/main/java/org/jabref/gui/importer/fetcher/WebSearchPaneViewModel.java

```diff
@@ -91,11 +111,67 @@ public void search() {
 
         BackgroundTask<ParserResult> task;
         task = BackgroundTask.wrap(() -> new ParserResult(activeFetcher.performSearch(getQuery().trim())))
-                             .withInitialMessage(Localization.lang("Processing %0", getQuery().trim()));
+                .withInitialMessage(Localization.lang("Processing %0", getQuery().trim()));
         task.onFailure(dialogService::showErrorDialogAndWait);
 
         ImportEntriesDialog dialog = new ImportEntriesDialog(stateManager.getActiveDatabase().get(), task);
         dialog.setTitle(activeFetcher.getName());
         dialogService.showCustomDialogAndWait(dialog);
     }
+
+    public void validateQueryStringAndGiveColorFeedback(TextField querySource, String queryString) {
+        if (!queryString.strip().isBlank() && !isPatternMatched(queryPattern, queryString)) {
+            if (isPatternMatched(laxQueryPattern, queryString)) {
+                setPseudoClassToUnsupported(querySource);
+                querySource.setTooltip(new Tooltip(Localization.lang("This query uses unsupported fields.")));
+            } else {
+                setPseudoClassToInvalid(querySource);
+                querySource.setTooltip(new Tooltip(Localization.lang("This query uses unsupported syntax.")));
+            }
+        } else if (containsYearAndYearRange(queryString)) {
+            setPseudoClassToInvalid(querySource);
+            querySource.setTooltip(new Tooltip(Localization.lang("The query cannot contain a year and year-range field.")));
+        } else {
+            setPseudoClassToValid(querySource);
+        }
+    }
+
+    private void setPseudoClassToUnsupported(TextField querySource) {
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("invalid"), false);
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("unsupported"), true);
+    }
+
+    public void setPseudoClassToValid(TextField querySource) {
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("invalid"), false);
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("unsupported"), false);
+    }
+
+    private void setPseudoClassToInvalid(TextField querySource) {
+        querySource.pseudoClassStateChanged(PseudoClass.getPseudoClass("invalid"), true);
+    }
+
+    private boolean containsYearAndYearRange(String queryString) {
+        return queryString.toLowerCase().contains("year:") && queryString.toLowerCase().contains("year-range:");
+    }
+
+    // Use Thread Pool to control the match time
+    private boolean isPatternMatched(Pattern pattern, String queryString) {
+        boolean isMatched;
+        Callable<Boolean> call = () -> {
```

Comment:
You better use our BackGroundTask.wrap(...)

Reply:
Done! Please check it!

Reply:
Yep, works better now.